generator/src/Parameter.php

```diff
@@ -21,49 +21,63 @@ public function __construct(\SimpleXMLElement $parameter, ?PhpStanFunction $phpS
     }
 
     /**
-     * Returns the type as declared in the doc.
-     * @return string
+     * Tries to identify the typehint from the doc-block parameter
      */
-    public function getType(): string
+    public function getSignatureType(): string
     {
-        $type = $this->parameter->type->__toString();
-        $strType = Type::toRootNamespace($type);
-        if ($strType !== 'mixed' && $strType !== 'resource' && $this->phpStanFunction !== null) {
-            $phpStanParameter = $this->phpStanFunction->getParameter($this->getParameter());
-            if ($phpStanParameter) {
-                // Let's make the parameter nullable if it is by reference and is used only for writing.
-                if ($phpStanParameter->isWriteOnly()) {
-                    $strType = '?'.$strType;
-                }
-            }
+        $coreType = $this->getDocBlockType();
+        //list all types in the doc-block
+        $types = explode('|', $coreType);
+        //no typehint exists for thoses cases
+        if (in_array('resource', $types) || in_array('mixed', $types)) {
+            return '';
+        }
+        //remove 'null' from the list to identify if the signature type should be nullable
+        $nullablePosition = array_search('null', $types);
+        if ($nullablePosition !== false) {
+            array_splice($types, $nullablePosition, 1);
+        }
+        if (count($types) === 0) {
+            throw new \RuntimeException('Error when trying to extract parameter type');
+        }
+        //if there is still several types, no typehint
+        if (count($types) > 1) {
+            return '';
+        }
+
+        $finalType = $types[0];
+        if (strpos($finalType, '[]') !== false) {
+            $finalType = 'array'; //generics cannot be typehinted and have to be turned into array
```

Comment:
```suggestion
            $finalType = 'iterable'; //generics cannot be typehinted and have to be turned into array
```

[] does not mean necessarily array, it could be a Traversable